### PR TITLE
Added new parameter 'sub_token_mode' to 'pretrained_transformer_mismatched_embedder' class to support first sub-token embedding (#4363)

### LINKED ISSUE #4363: Word embedding in BERT can be the first/last subtoken embedding?
In v0.9.0, when we use BERT, the contextualized word embedding is set to be the first (or last) subtoken embedding (in `BertEmbedder`).

However, in v1.0.0, the contextualized word embedding is set to be the average of all the embeddings of the subtokens in a word (in `PretrainedTransformerMismatchedEmbedder`).

The former treatment (v0.9.0) is adopted in the original BERT paper, and so this option will be added again in the future?

Thanks.

### PR DESCRIPTION
**Pull request by prajaktakini-vmware**
Fixes #4363 .

Changes proposed in this pull request:
Currently 'PretrainedTransformerMismatchedEmbedder' in allennlp only allows to return average of embeddings of all
sub-tokens of a word whereas Original BERT Model suggests to take first/last sub-token embedding.
Added the flexibility to choose first subtoken embedding based on a parameter `sub_token_mode`

- If 'sub_token_mode' is set to 'first', return first sub-token representation as word-level representation
- If 'sub_token_mode' is set to 'avg', return average of all the sub-tokens representation as word-level representation
- If 'sub_token_mode' is not specified it defaults to 'avg'
- If invalid 'sub_token_mode' is provided, throw 'ConfigurationError'

Acknowledgement of support by  @pab-vmware, @annajung, @agururajvais, @divijbajaj

### CODE REVIEW

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py [hunk 1]
```diff
@@ -111,15 +119,53 @@ def forward(
...
+            orig_embeddings = span_embeddings_sum / torch.clamp_min(span_embeddings_len, 1)
+
+            # All the places where the span length is zero, write in zeros.
+            orig_embeddings[(span_embeddings_len == 0).expand(orig_embeddings.shape)] = 0
 
-        # All the places where the span length is zero, write in zeros.
-        orig_embeddings[(span_embeddings_len == 0).expand(orig_embeddings.shape)] = 0
+        # If invalid "sub_token_mode" is provided, throw error
+        else:
+            raise ValueError("Do not recognise 'sub_token_mode' {}".format(self.sub_token_mode))
```
Comment (dirkgr) [inline allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:169]: This should be a `ConfigurationError`.

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py [hunk 2]
```diff
@@ -111,15 +119,53 @@ def forward(
...
+        # If "sub_token_mode" is set to "first", return the first sub-token embedding
+        if self.sub_token_mode == "first":
+            # Shape: (batch_size, num_orig_tokens, max_span_length)
+            span_mask_bool = torch.zeros(span_mask.shape, dtype=torch.bool)
+
+            # Shape: 1-dimensional tensor
+            max_span_range_indices = torch.arange(max_span_length, dtype=torch.long).view(1, 1, -1)
+
+            # Shape (batch_size, num_orig_tokens, max_span_length)
+            span_mask = max_span_range_indices <= span_mask_bool
```
Comment (dirkgr) [inline allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:135]: What's going on here? `span_mask_bool` is always 0 at this point.

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py [hunk 3]
```diff
@@ -111,15 +119,53 @@ def forward(
...
+            span_mask = span_mask.unsqueeze(-1)
+
+            # Shape: (batch_size, num_orig_tokens, max_span_length, embedding_size)
+            span_embeddings *= span_mask
+
+            # Sum over embeddings of all sub-tokens of a word where except first sub-token,
+            # rest all sub-token's embeddings is zero padded
+            # Shape: (batch_size, num_orig_tokens, embedding_size)
+            orig_embeddings = span_embeddings.sum(2)
```
Comment (dirkgr) [inline allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:144]: I don't have the complete overview over this code, but doesn't this boil down to `orig_embeddings = span_embeddings[..., 0]`?

I might be missing a dimension somewhere.

**Review by dirkgr — CHANGES_REQUESTED** [review]
I might be missing something, but I think this can be simplified a bit.

Thanks for taking this on!

**Comment by prajaktakini-vmware** [comment]
> I might be missing something, but I think this can be simplified a bit.
> 
> Thanks for taking this on!

Hi Dirk, Yes this can be simplified with indexing based approach. Here We can get first subtoken embedding using indexing over the span embedding after applying zero padding on them. 
`span_embeddings *= span_mask  # zero out padding`
`span_embeddings[:,:,0, :]`


But to keep it similar to Masking based approach which has been used in `pretrained_transformer_mismatched_embedder` , I make it that way. 
Please suggest if the above indexing approach looks okay to you. Based on your suggesting, I will make changes to PR

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py (hunk 2 above)
Comment (prajaktakini-vmware) [reply allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:135]: yes, I am constructing the `span_mask` tensor here. Steps are as follows

1. Construct Boolean tensor with the dimension as `(batch_size, num_orig_tokens, max_span_length)`
2. (At line #135) Construct `span_mask` tensor by keeping first tensor from 3rd dimension i.e. `max_span_length`, as Boolean `True` and rest as Boolean `False`. This is done so that when multiplied with `span_embeddings`, it will only select first sub-token embedding and rest all will be zeroed.

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py (hunk 3 above)
Comment (prajaktakini-vmware) [reply allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:144]: Here `span_embeddings` is of shape `(batch_size, num_orig_tokens, max_span_length, embedding_size)` where from the 3rd dimension i.e. `max_span_length`, only first tensor has actual values (first subtoken embedding), rest have zero elements. 

When sum is taken along 3rd dimension i.e. `max_span_length` that will boil down to shape `(batch_size, num_orig_tokens, embedding_size)` and hence one of the dimension will be squeezed automatically by sum function

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py (hunk 1 above)
Comment (prajaktakini-vmware) [reply allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:169]: Thanks! will update it

File: allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py (hunk 3 above)
Comment (dirkgr) [reply allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:144]: So then `orig_embeddings = span_embeddings[..., 0, :]`, right? Why mess around with `span_mask` and `span_mask_bool` and all that?

Comment (dirkgr) [reply allennlp/modules/token_embedders/pretrained_transformer_mismatched_embedder.py:144]: Oh wait, you already answered that below. I should read ahead :-)

**Comment by dirkgr** [comment]
Yes, I think the indexing based approach is easier to understand, and it will run faster.

**Comment by prajaktakini-vmware** [comment]
> Yes, I think the indexing based approach is easier to understand, and it will run faster.

Hi @dirkgr, I have updated the PR with an indexing based approach to get the first sub-token embedding. Thanks!

**Review by dirkgr — APPROVED** [review]
This is great! Thank you also for the improvements in documentation.
